compliance/collection/main.go

```diff
@@ -182,17 +217,106 @@ func manageNodeScanLoop(ctx context.Context, rescanInterval time.Duration, scann
 	return sensorC
 }
 
-func scanNode(nodeName string, scanner nodeinventorizer.NodeInventorizer) (*sensor.MsgFromCompliance, error) {
-	result, err := scanner.Scan(nodeName)
+func runCachedScan(opts *cachedScanOpts) (*storage.NodeInventory, error) {
+	inventory, err := opts.scanner.Scan(opts.nodeName)
 	if err != nil {
 		return nil, err
 	}
+	inv, err := proto.Marshal(inventory)
+	if err != nil {
+		return nil, err
+	}
+	if err := os.WriteFile(fmt.Sprintf("%s/last_scan", opts.inventoryCachePath), inv, 0600); err != nil {
+		return nil, err
+	}
+
+	return inventory, nil
+}
+
+func createAndObserveMessage(nodeName string, inventory *storage.NodeInventory) *sensor.MsgFromCompliance {
 	msg := &sensor.MsgFromCompliance{
 		Node: nodeName,
-		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: result},
+		Msg:  &sensor.MsgFromCompliance_NodeInventory{NodeInventory: inventory},
 	}
 	cmetrics.ObserveInventoryProtobufMessage(msg)
-	return msg, nil
+	return msg
+}
+
+// scanNodeWithBackoff runs scans with a linear backoff based on a file to not overstrain a Node if the container keeps restarting.
+// The backoff file will only be encountered if the previous container is killed during a call to scanNodeWithBackoff.
+// Note: This does not prevent strain in case of repeated pod recreation, as it is based on an EmptyDir.
+func scanNodeWithBackoff(opts *cachedScanOpts) (*sensor.MsgFromCompliance, error) {
+	backoffIntervalSeconds := int64(env.NodeInventoryInitialBackoff.DurationSetting() / time.Second)
+
+	backoffFile, err := os.ReadFile(fmt.Sprintf("%s/backoff", opts.inventoryCachePath))
+	defer func() {
+		if err := os.Remove(fmt.Sprintf("%s/backoff", opts.inventoryCachePath)); err != nil {
+			log.Warnf("Could not remove backoff state file: %v", err)
+		}
+	}()
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			log.Debug("No backoff found, continuing without pause")
+		} else {
+			return nil, err
+		}
+	} else {
+		// We have an existing backoff counter
+		backoffIntervalMillis, err := strconv.ParseInt(string(backoffFile), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+		backoffIntervalSeconds = int64(time.Duration(backoffIntervalMillis) / time.Second)
+		log.Infof("Found existing backoff. Waiting before running next inventory")
```

Comment:
I'd say this should be a warning. We should tell the user that the previous scan did not complete successfully and that this time we'll delay the scan in order not to overload the node.
It would be nice to also tell for how long we're delaying this scan. Which hints that logging here and logging in `waitAndIncreaseBackoff` is double-logging and can be reduced to just one.

Reply:
Done!
I'll revisit the logging and call stack when reorganising in the next commits.